### DISCUSSION TITLE
EmitCompilerGeneratedFiles while output redirected back into project folder causes duplicate errors

### DISCUSSION THREAD

**Discussion by chismar** [dotnet/roslyn#72977] [Q&A]
I'm migrating from an old precursor of Source Generators by [AArnott](https://github.com/AArnott) which I have customized a bit, to the  modern Source Generators.

In my previous implementation Source Generators outputted generated code on disk which then passed compilation like normal file. It was also completely intellisensed, source controlled and debuggable.
If I'd remove the generator nothing would change as the generated code has already been created.

After migration I have lost all those QoL features like the code being properly intellisensed and especially DEBUGGABLE.
I've also placed the files adjacent to the sources. So if I had Test.cs then there'd be a Test.generated.cs next to it.
There is a similar discussion https://github.com/dotnet/roslyn/discussions/64705, but I feel like the answer is not something I can accept.

Whatever files are outputted are unable to be matched by Rider, VS or otherwise to the symbols in the assembly, since they are not the ones who are fed to the compiler.

Instead the compiler is fed by the inmemory generated sources, while the emitted ones are just for show.

This is very frustrating, and I want to somehow get my QoL features back.

Prefferably source generator would emit the source files and those would be compiled, instead of those in memory. 
Such that if I'd remove the source generator as a dependency the sources would remain.
And such that if I rebuild the project with the generator there'd be no duplication errors, since only the emitted files are used.
And such that I could place a breakpoint in those files and actually hit it.

Is there anything I can do for this to work? 
Because the entire point of generating human readable sources is lost if I can't actually hit the debug breakpoints in it. 

I mean, those are SOURCE generators after all, not inmemorytexts generators. One would assume you'd want the actual source files as the output, with all the features associated.

**Comment by CyrusNajmabadi** [comment]
Why did you move off of the old system?  

> After migration I have lost all those QoL features like the code being properly intellisensed

Can you explain what you mean by this?  Intellisense should work with SG files just fine.  If you're not seeing that, can you file a new issue with repro steps so we can try out your scenario?

> Such that if I'd remove the source generator as a dependency the sources would remain.

This is doable.  You can ask the compiler to emit the SG files to disk, then remove your generator, then only reference the generated files.

I would actually recommend this as well, as it will vastly reduce the complexity in the system here.

**Reply by chismar** [reply]
> Why did you move off of the old system?

Because source generators are also supported in Unity, while a custom target in a csproj isn't, thus new source generators are the only way to create code both in Unity and a separate csproj. 
Also supporting a custom infrastructure for source generators for 7 years is simply tiring. 

>Can you explain what you mean by this? Intellisense should work with SG files just fine. If you're not seeing that, can you file a new issue with repro steps so we can try out your scenario?

It's not "not working at all" per se, more like it breaks, since the emitted files are copies of the inmemory ones and that confuses the hell out of rider and vs. Akin to how dotnet can't build them due to duplicates if you don't do "Compile Remove="*.g.cs""

And none of the code emitted or those found in temp files, shown by Rider for example, the ones you get to when look for definition, are mapped to symbols in the pdb, which means I can't use them for debugging.

>I would actually recommend this as well, as it will vastly reduce the complexity in the system here.

I need the generator to run regularly to update the datastructures I generate. I generate a lot of boilerplate for a custom game networking framework, all the sync methods, structures and other plumbing.



Basically, my issue is the generator emits new sources as some inmemory versions, embeds them into a Compilation and goes on.
But what I need, preferrably, is to emit files on disk whenever SG is run and add THOSE to the Compilation, so that everything is mapped as if were written in a normal way.

I could just write the files on disk directly instead of calling AddSource, but then compiler doesn't know of them, of course.
I could emit the files and then call AddSource with a rooted hint just like the ones on disk, but there is code that prevents rooted hintNames and then embeds sources into pdb anyway.

At this point I'm working on a solution that takes Harmony IL patching library, loads it with the Source Generator and then patch the hell out of compiler code so that it actually takes rooted syntaxtree paths and wouldn't embed them into pdbs.
EDIT: IL patching failed as soon as I tried - no IL Emit supported. I'm at a loss at this point

As you might imagine that's not ideal.


So the ROOT of my frustration is the fact that I can't set a breakpoint in my IDE and actually hit it, since the symbols have no mention of the emitted files or the temp files. I'd probably live with the fact that my sources are not proper sources, but not having proper symbols is appalling

**Reply by chismar** [reply]
uh oh, looks like newer visual studio actually hits debug the source generated files
maybe if I update Rider too it will work?

still, the fact that source generators couldn't just place sources like normal files and instead rely on inmemory stuff... just why? Very counter-intuitive, not to mention requires additional support from IDEs

**Reply by chismar** [reply]
na ah, new Rider wouldn't hit breakpoints in source generated files
yeah, cool, generated source files that can only be debugged in latest visual studio with source files placed not at source folder but wherever
something on the level of C++ toolchain pain, not something I would ever imagine a .net developer would experience after all this years

**Reply by chismar** [reply]
Same behavior with VS and Unity, can't hit a breakpoint and debugger in general goes bonkers when trying to stepthrough

**Reply by chismar** [reply]
Not only that but it also caches something and thus the executed code doesn't even match the latest generated source

I guess at this point I just have to give up and abandon SGs? For Unity I suppose I'll just create shims and #if_def generated code, then redirect shared code to look for sources in assets folder.

Maybe there's a way to force running analyzer before compile and just dump sources with File.WriteAllText? 

Never could I have imagined Source Generators to be so badly implemented, completely useless. They should've just generated sources and not require support from every IDE in different configurations.

**Reply by CyrusNajmabadi** [reply]
> Because source generators are also supported in Unity, while a custom target in a csproj isn't,

That sounds like a unity issue.  Anything new we did in this space could have the same unity restrictions, so we can't design new work just to workaround potential problems in other tools you are using. :)

> It's not "not working at all" per se, more like it breaks, since the emitted files are copies of the inmemory ones and that confuses the hell out of rider and vs. 

Any rider issues should be reported to them.  They do not use Roslyn for this, so bugs there are things they need to fix.

> Akin to how dotnet can't build them due to duplicates if you don't do "Compile Remove="*.g.cs""

My point was: tell the compiler to emit htem, then just reference the emitted files (and not the generator) anymore.  There should be no problems with 'duplicates' then.

**Reply by CyrusNajmabadi** [reply]
> But what I need, preferrably, is to emit files on disk whenever SG is run and add THOSE to the Compilation, so that everything is mapped as if were written in a normal way.

There's nothing that would make that system less buggy than what you're encountering now.  If you have issues now, then you'd have issues with the above, because that's basically exactly how SGs work today, just eliminating a complex intermediary IO step that could cause lots of issues. :)

> So the ROOT of my frustration is the fact that I can't set a breakpoint in my IDE and actually hit it,

I cannot repro this at all.  It sounds like a bug in your setup (possibly your generator).  If you can attach a repro case, we can look into what's going on here.  Absent that, it will be hard to make progress on this as you're reporting things that may be specific to your setup, along with unity and your SG.

> still, the fact that source generators couldn't just place sources like normal files and instead rely on inmemory stuff... just why?

Because it's *much* simpler.  We remove an unnecessary indirection, along with avoiding the disk, which is a huge source of  issues commonly during building (Defender alone is a massive pain for things like this).

> Very counter-intuitive, not to mention requires additional support from IDEs

Additional support would be required no matter what.  We'd need all the same support as today, just with disk support instead of in-memory support.  In-memory-support simplifies a ton as we don't have to deal with any disk issues at all as we can literally just use the SourceText your generator generated.  

> na ah, new Rider wouldn't hit breakpoints in source generated files

That sounds like a rider issue.  You should report this to them.  We don't change our designs around external tools, esp. as those external tools may have the exact same set of issues (or worse) if we change this sort of thing.

**Reply by CyrusNajmabadi** [reply]
> still, the fact that source generators couldn't just place sources like normal files and instead rely on inmemory stuff... just why?

Because it's *much* simpler.  We remove an unnecessary indirection, along with avoiding the disk, which is a huge source of  issues commonly during building (Defender alone is a massive pain for things like this).

> Very counter-intuitive, not to mention requires additional support from IDEs

Additional support would be required no matter what.  We'd need all the same support as today, just with disk support instead of in-memory support.  In-memory-support simplifies a ton as we don't have to deal with any disk issues at all as we can literally just use the SourceText your generator generated.  

> Same behavior with VS and Unity, can't hit a breakpoint and debugger in general goes bonkers when trying to stepthrough

Sounds like a unity issue.  If you can report this to them, or provide us with a repro case, we could take a look.  A repro will be needed here as you're the only customer reporting this, so it's likely something to do with your local setup.

**Reply by CyrusNajmabadi** [reply]
> Never could I have imagined Source Generators to be so badly implemented, completely useless. They should've just generated sources and not require support from every IDE in different configurations.

Source generators would always require support from whatever host you're using.  That's the point of them.  They run *within* a compilation, not as a pre-step.  If you want a generation pre-step system, then it's trivial to build such a thing, and many already exist for all sorts of generation tasks.  They work exactly like you describe, emitting normal c# files that are treated by all downstream systems as normal files.  The point of SGs was to create a system that enabled something those prior systems did not support at all.  So, if you don't need that (which it soundsl ike you don't), it sounds like you're using the wrogn tool for the job.

During that, it sounds like you may have run into bugs related to your setup (not uncommon, given how complex setups can be).  For rider issues, please let them know.  For VS issues, please provide us with a repro so we can take a look.

Thanks! :)

**Reply by chismar** [reply]
My apologies for any antagonizing sounding remarks.
I sleep very little, am very frustrated at the SG implementation, and have very tight deadlines. 

On the topic of "other systems" - SGs are completely capable of supporting my usecase were they not aritificially limited.

Why only support the case of embedded sources?
Why force specific layout of generated files?

Why can't I call AddSource with a rooted file path, save said file immediately at said file (somewhere in the ProjectDir) and expect it to just work? All of that is simply a matter of like 2 IFs in the compiler source.

One around GenerateAndEmit where it looks for generated sources and only assumes those are inmemory ones
One inside the generators source where it forcibly applies prefixes to the hintpath and doesn't allow rooted paths
And a line to override the input syntaxTree if source generator has a newer version for it (aka if it generated a file with the same path as one of the input syntax trees)

It would work like a charm, zero IDE support (from the point of view of IDE there's only normal source files in the project, no weird temp files and such)

**Reply by chismar** [reply]
By that point I abandoned idea of migrating to SGs, since even if I convince anyone that such changes are necessary, Unity won't update its code for the forseable future.

Went back to my custom generator.

**Reply by chismar** [reply]
On the repro - I'd come around to that at some point in the future maybe, but as of now I need like 30 hours of daylight per day or a divine intervention to make everything in time. No time for repros.

**Reply by CyrusNajmabadi** [reply]
>No time for repros.

Unfortunately, there's not a lot we can do here then.  :(

**Reply by CyrusNajmabadi** [reply]
> Why only support the case of embedded sources?

Because that's the use case they were intended to solve.  Non-embedded sources already have lots of other tools that can support that case.

> Why force specific layout of generated files?

Not sure what you mean by this.

> Why can't I call AddSource with a rooted file path, save said file immediately at said file (somewhere in the ProjectDir) and expect it to just work?

It very much *is not*.  You've now introduced untracked IO into the system.  Which means things like incrementality break completely.  Tools like msbuild would have no idea about these files

>  All of that is simply a matter of like 2 IFs in the compiler source.

I honestly recommend you try this.  You'll see that thsi doesn't solve things.  The IDE code, for example, could add that `if` to save out the files, but you'd still see all the same issues you're running into today with tooling.  Why?  Because none of the tooling woudl have any clue what to do with those files (as they would not be part of any project).  We could do all the work to add it to the project.  But that's what teh code is *already doing* today.

**Reply by CyrusNajmabadi** [reply]
> It would work like a charm, zero IDE support (from the point of view of IDE there's only normal source files in the project, no weird temp files and such)

THis is not the case.  you say "in the project", but this would not be "in the project" as the project doesn't reference that file *at all*.   We'd have to do the work to get the generated file actually into the project system.  Which is exactly what the SG infrastructure does today, just without intermediary IO ops.

**Reply by CyrusNajmabadi** [reply]
> Unity won't update its code for the forseable future.

I'm going to close this discussion out.  My guess is that there is something about your unity setup that is causing problems.  So far, we haven't heard from anyone else about any of these issues, so it's likely a mis-authoring, or some other bug in the tools/tasks there.  If you want to file a repro in the future, we'd be able to look at the logs to see what's up.  Thanks!
